extensions/ql-vscode/src/common/vscode/file-path-discovery.ts

```diff
@@ -0,0 +1,190 @@
+   * @param fileWatchPattern Passed to `vscode.RelativePattern` to determine the files to watch for changes to.
+   */
+  constructor(name: string, private readonly fileWatchPattern: string) {
+    super(name, extLogger);
+
+    this.onDidChangePathsEmitter = this.push(new EventEmitter<void>());
+    this.push(
+      workspace.onDidChangeWorkspaceFolders(
+        this.workspaceFoldersChanged.bind(this),
+      ),
+    );
+    this.push(this.watcher.onDidChange(this.fileChanged.bind(this)));
+  }
+
+  /**
+   * Compute any extra data to be stored regarding the given path.
+   */
+  protected abstract getDataForPath(path: string): Promise<T>;
+
+  /**
+   * Is the given path relevant to this discovery operation?
+   */
+  protected abstract pathIsRelevant(path: string): boolean;
+
+  /**
+   * Should the given new data overwrite the existing data we have stored?
+   */
+  protected abstract shouldOverwriteExistingData(
+    newData: T,
+    existingData: T,
+  ): boolean;
+
+  /**
+   * Do the initial scan of the entire workspace and set up watchers for future changes.
+   */
+  public async initialRefresh() {
+    getOnDiskWorkspaceFoldersObjects().forEach((workspaceFolder) => {
+      this.changedFilePaths.addPath(workspaceFolder.uri.fsPath);
+    });
+
+    this.updateWatchers();
+    return this.refresh();
+  }
+
+  private workspaceFoldersChanged(event: WorkspaceFoldersChangeEvent) {
+    event.added.forEach((workspaceFolder) => {
+      this.changedFilePaths.addPath(workspaceFolder.uri.fsPath);
+    });
+    event.removed.forEach((workspaceFolder) => {
+      this.changedFilePaths.addPath(workspaceFolder.uri.fsPath);
+    });
+
+    this.updateWatchers();
+    void this.refresh();
+  }
+
+  private updateWatchers() {
+    this.watcher.clear();
+    for (const workspaceFolder of getOnDiskWorkspaceFoldersObjects()) {
+      // Watch for changes to individual files
+      this.watcher.addWatch(
+        new RelativePattern(workspaceFolder, this.fileWatchPattern),
+      );
+      // need to explicitly watch for changes to directories themselves.
+      this.watcher.addWatch(new RelativePattern(workspaceFolder, "**/"));
+    }
+  }
+
+  private fileChanged(uri: Uri) {
+    this.changedFilePaths.addPath(uri.fsPath);
+    void this.refresh();
+  }
+
+  protected async discover() {
+    let pathsUpdated = false;
+    let path: string | undefined;
+    while ((path = this.changedFilePaths.popPath()) !== undefined) {
+      if (await this.handledChangedPath(path)) {
+        pathsUpdated = true;
+      }
+    }
+
+    if (pathsUpdated) {
+      this.onDidChangePathsEmitter.fire();
+    }
+  }
+
+  private async handledChangedPath(path: string): Promise<boolean> {
+    if (!(await exists(path)) || !this.pathIsInWorkspace(path)) {
+      return this.handledRemovedPath(path);
+    }
+    if ((await lstat(path)).isDirectory()) {
+      return await this.handleChangedDirectory(path);
+    }
```

Comment:
Are we able to combine the `exists` and `lstat` calls? I believe these will make very similar syscalls, and you should be able to derive if a file doesn't exist from the error returned by `lstat`. The example given in [the Node.js docs](https://nodejs.org/api/fs.html#fsexistspath-callback) is:

```js
  if (err) {
    if (err.code === 'ENOENT') {
      console.error('myfile does not exist');
      return;
    }

    throw err;
  }
```

Reply:
I'm not sure what's best. I agree you could just call `lstat` and then catch the error if the file doesn't exist. I did consider something like this but it doesn't seem great to use exceptions in this way. I'll have more of a think on it.

I'm also now realising there's a lot of "time-of-check-time-of-read" bugs here, where we check the path exists and then do stuff with it a bit later on. I'll have a think about how severe these are too. It's mostly ok if the discovery process fails so long as we can recover from it and try again and don't corrupt any data in the process.